### DISCUSSION TITLE
테스트, 잘 되어 가나요?

### DISCUSSION THREAD

**Discussion by woowapark** [woowacourse/react-subway-map#4] [General]
- 어떤 테스트 도구들을 사용하고 있나요?
- 테스트를 고려하기 위해 어떤 방식으로 개발하고 있나요? (TDD를 시도해본다, 우선 구현해본 다음 테스트를 추가하면 좋을 것 같은 부분을 찾아본다, 테스트 시나리오부터 써본다 등등) 
- 테스트가 개발하는 데 도움이 되는 것 같나요?
- 테스트를 고려하는 과정에서 어려움이 있다면 어떤 어려움이 있고, 어떻게 해결하고 있나요?
- 그 외 나누고 싶은 이야기, 경험

**Comment by yujo11** [comment]
### 파노 그루밍 - RTL

**선택 이유**

- E2E는 번거로워서 RTL을 선택. 컴포넌트만 바로 테스트 할 수 있어서

**어려웠던 점**

- 회원가입 폼이 회원가입에 대한 기능까지 담당해야 된다고 생각해서 BDD로 로그인까지 테스트를 하려고 했으나 잘 안 됐다.

---

### 피터 서니 - Storybook, RTL(예정)

- 기능 구현을 우선시 하느라 아직 못 정함.

---

### 유조 - Stroybook, RTL(예정)

- 기능 구현을 우선시 하느라 아직 못 정함.

**Comment by 2SOOY** [comment]
### 곤이
- `스토리북` UI 테스트 + `saga test plan` 사용
- RTL을 도입하는데 비용이 많이 들 것이라고 예상
- E2E와 같은 전반적인 컴포넌트에 대한 테스트 보다는 리듀서 단위의 테스트를 진행할 예정

### 지그
- RTL or Cypress를 도입할 것
    - RTL을 사용하는 것 자체 러닝커브가 높다고 판단
    - 리액트에서도 Cypress 테스트가 가능하다는 것을 파악하여 RTL에 비해 러닝커브가 비교적 낮은 Cypress를 통한 E2E 테스트 진행 
- UI 테스트 - 스토리북 테스트

### 동동
- RTL을 적용할 것 + UNIT TEST
   - E2E 테스트 보다는 테스트 범위를 축소할 예정 - 테스트 자체도 비용이 높기 때문에   
   - 컴포넌트 개별단위의 유닛 테스트를 진행할 것 

### 도비
- RTL 사용
- 테스트 사이클을 돌기 어려움 (실패 → 기능 → 성공)
- 로그인 / 회원가입 - 정상적인 INPUT에 대한 페이지 이동 / 버튼 활성화 등 진행함
- 기술적으로 어려웠던 점
    - RTL 사용할 때, 함수 mocking 하는 부분 ex) uesSelector / useDispatch
        - 어디까지 mocking해야할까 ?? 에 대한 고민
        - mocking에 의해 함수간 의존성 때문에 어플리케이션이 동작이 망가지기도 함
    - mocking이란?
        - 실제 관리하는 상태의 변화는 일으키지 않고 우리가 원하는 상태를 얻었을 때의 예상하는 동작을 테스트
        `참고` [프론트엔드 TDD](https://www.youtube.com/watch?v=L1dtkLeIz-M)

### 콜린
- RTL + JEST를 사용함
- 이전 미션에는 실제 useDispatch와 useSelector등을 사용하여 상태가  계속 변경되어 1회성 테스트여서 효용이 없다고 생각
- 기존의 상태를 건드리지 않도록 함수를 mocking하는 작업을 진행함 (useDispatch, useSelector)
- 어디까지, 무엇을 mocking 해야할지 고민이 필요

**Comment by swon3210** [comment]
### **[ 카일 ]**

RTL 진행을 했지만, 비동기 테스트에 있어서는 애를 많이 먹고 있음. 더 해봐야 효용이나 어려움을 제대로 알 수 있을 듯. 기능 구현 먼저 하는 것은 다른 크루들과 비슷

### **[ 인치 ]**

구현하고 테스트 작성하면 리팩토링 할때 주요 기능들을 다시 검증해야 하는 불편함이 감소됨 <- 이 정도를 본인 수준에 맞춰서 적용하고 있음

### **[ 티케 ]**

TDD 를 하려고 했지만 테스트 커버리지를 크게 잡는 건 불가능 해서 관리 페이지 위주로 잡고 있다. API 테스트를 진행할 때 이를 일일히 확인하기 어렵기 때문에 이 부분은 확실히 테스트가 도움이 되는 듯. 그러나 E2E 는 RTL 로 하는게 확실히 어렵다.

### **[ 썬 ]**

RTL, Storybook 사용 중. RTL을 아직 아예 사용해본적이 없음. 스토리북은 잘 쓰고 있는데 bottom up 방식에 필수적이라고 생각.

### **[ 크리스 ]**

E2E 를 하고 싶지만 RTL 이 이를 하는데 너무 불편하게 되어 있음. 테스트 에러가 너무 많아서 진척이 너무 느렸기 때문에 일단 기능 구현을 먼저하고 있음

테스트 도구를 능숙하게 사용할 수 있게 되면 분명 도움이 될 것이라 생각하지만, 도구 자체의 사용법을 하나씩 배워야 하는 단계에서는 많이 도움이 되지는 못할 것이라 생각(비용적인 측면에서)

테스트 도구 자체를 쓰는 법을 배우는 시간을 따로 내야 한다는 점이 가장 힘들다
